crates/config/src/adkg.rs

```diff
@@ -0,0 +1,147 @@
+use crate::keys::libp2p_keypair_serde;
+use anyhow::{Context, anyhow};
+use itertools::Itertools;
+use libp2p::{Multiaddr, PeerId, identity};
+use serde::{Deserialize, Serialize};
+use std::num::NonZeroUsize;
+use std::str::FromStr;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct NodeDetail {
+    pub id: NonZeroUsize,
+    #[serde(flatten)]
+    pub public_key_material: PublicKeyMaterial,
+    pub multiaddr: Multiaddr,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct GroupConfig {
+    pub n: NonZeroUsize,
+    pub t: NonZeroUsize,
+    pub t_reconstruction: NonZeroUsize,
+    pub start_time: chrono::DateTime<chrono::Utc>,
+    pub nodes: Vec<NodeDetail>,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct AdkgSecret {
+    pub adkg_scheme_name: String,
+    pub genesis_timestamp: i64,
+    pub sk: String,
+}
+
+/// The ADKG public output on a source group (i.e., the one used by the ADKG), and on a destination
+/// group (i.e., the one used for signatures, etc.)
+#[derive(Clone, Serialize, Deserialize)]
+pub struct AdkgPublic {
+    pub adkg_scheme_name: String,
+    pub genesis_timestamp: i64,
+
+    /// The destination group public key used for signatures
+    pub group_pk: String,
+    /// The destination node public keys used for signatures
+    pub node_pks: Vec<AdkgNodePk>,
+
+    /// The group public key obtained in the ADKG source group.
+    #[serde(default)]
+    pub group_pk_source: String,
+    /// The node public keys obtained in the ADKG source group.
+    #[serde(default)]
+    pub node_pks_source: Vec<AdkgNodePk>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct AdkgNodePk {
+    pub id: NonZeroUsize,
+    pub pk: String,
+    pub peer_id: PeerId,
+    pub multiaddr: Multiaddr,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct PrivateKeyMaterial {
+    pub adkg_sk: String,
+    #[serde(with = "libp2p_keypair_serde")]
+    pub libp2p_sk: identity::Keypair,
```

Comment:
we could replace the type with `Libp2pKeyWrapper`, and remove the `mod libp2p_keypair_serde`, we'll just need to slightly change how the key is accessed.